### DISCUSSION TITLE
PathBase vs Prefix (HttpSys)

### DISCUSSION THREAD

**Discussion by whoaskedfrfr** [dotnet/aspnetcore#57139] [General]
Hey! I'm confused about what's the difference of prefix and path base.

Recently I made a change in a code base from:

```
webBuilder.UseUrls("https://*:888/api");
```

to

```
webBuilder.UseUrls("https://*:888");

/* ... */

appBuilder.UsePathBase("/api");
```

Everything worked fine until it reached an environment with other apps making it crash with `The prefix 'https://*:888/' is already registered.`

I was under the impression the same thing. Can someone explain to me what are the differences?

**Comment by engineering87** [comment]
Hi there,
**UseUrls** method specifies the URLs on which Kestrel server should listen for incoming HTTP requests and combines the protocol, host, port and path prefix (`/api`) to define where the server listens for requests.
In your case:

```charsp
webBuilder.UseUrls("https://*:888/api");
```
restricts the server to handle requests starting with `/api` on port 888.

**UsePathBase** method modifies the request path that the application sees and allowing the application to act as if it's hosted at a subpath. It doesn’t affect where the server listens (like UseUrls) but changes how the requests are routed internally within the application.

By removing `/api` from **UseUrls**, your application began listening on `https://*:888/`, which is more general than `https://*:888/api`. If another application is already registered to listen on "https://*:888/", it causes a conflict, resulting in the error: `"The prefix 'https://*:888/' is already registered."`

So to solve the issue, you have to assign different ports for each application to avoid conflicts or use a reverse proxy to route requests to different applications based on the path.
